python/mlx/nn/layers/recurrent.py

```diff
@@ -0,0 +1,262 @@
+        self.br = _bias_init(hidden_size, bias)
+        # n
+        self.Wxn = _weight_init(input_size, hidden_size)
+        self.Whn = _weight_init(hidden_size, hidden_size)
+        self.bn = _bias_init(hidden_size, bias)
+        self.bhn = _bias_init(hidden_size, bias)
+        # z
+        self.Wxz = _weight_init(input_size, hidden_size)
+        self.Whz = _weight_init(hidden_size, hidden_size)
+        self.bz = _bias_init(hidden_size, bias)
+
+    def _extra_repr(self):
+        return f"input_dims={self.Wxr.shape[0]}, hidden_size={self.Wxr.shape[-1]}, bias={self.bias}"
+
+    def __call__(self, x):
+        x_proj_r = x @ self.Wxr
+        x_proj_n = x @ self.Wxn
+        x_proj_z = x @ self.Wxz
+
+        if self.bias:
+            x_proj_r += self.br
+            x_proj_n += self.bn
+            x_proj_z += self.bz
+
+        all_hidden = []
+        curr_hidden = mx.zeros(shape=(self.Whr.shape[0],))
+
+        for idx in range(x.shape[-2]):
+            r = x_proj_r[..., idx, :] + curr_hidden @ self.Whr
+            r = mx.sigmoid(r)
+
+            z = x_proj_z[..., idx, :] + curr_hidden @ self.Whz
+            z = mx.sigmoid(z)
+
+            n_x = x_proj_n[..., idx, :]
+
+            n_h = curr_hidden @ self.Whn
+            if self.bias:
+                n_h += self.bhn
+
+            n = n_x + r * n_h
+            n = mx.tanh(n)
+
+            curr_hidden = (1 - z) * n + z * curr_hidden
+            all_hidden.append(curr_hidden)
+
+        return mx.stack(all_hidden, axis=-2)
+
+
+class LSTM(Module):
+    r"""Apply one LSTM recurrent unit to a sequence of
+    shape ``NLD`` or ``LD``, where:
+        - ``N`` is the optional batch dimension
+        - ``L`` is the sequence length
+        - ``D`` is the input's feature dimension
+
+    Concretely, for each element of the sequence, computes:
+
+    .. math::
+        i_t = \sigma (W_{xi}x_t + b_{xi} + W_{hi}h_t + b_{hi})
+        f_t = \sigma (W_{xf}x_t + b_{xf} + W_{hf}h_t + b_{hf})
+        g_t = \text{tanh} (W_{xg}x_t + b_{xg} + W_{hg}h_t + b_{hg})
+        o_t = \sigma (W_{xo}x_t + b_{xo} + W_{ho}h_t + b_{ho})
+        c_{t + 1} = f_t \odot c_t + i_t \odot g_t
+        h_{t + 1} = o_t \text{tanh}(c_{t + 1})
+
+    The hidden state :math:`h` and cell state :math:`c` have shape ``NH``
+    or ``H``, depending whether the input is batched or not. Returns two
+    arrays: the hidden state and the cell state at each time step, both
+    of shape ``NLH`` or ``LH``.
+
+    Args:
+        input_size (int): Dimension of the input :math:`x`
+        hidden_size (int): Dimension of the hidden state :math:`h`
+        bias (bool): Whether to use biases or not. Default: `True`.
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        bias: bool = True,
+    ):
+        super().__init__()
+        self.bias = bias
+        # i
+        self.Wxi = _weight_init(input_size, hidden_size)
```

Comment:
Should we collapse Wxi ,Whi, Wxf, Wxo into one matrix? It’s more efficient to do 1 matmul instead of 4.

I’m also happy to help with this code if you want to collaborate, I’m working on models that require an LSTM.

Reply:
Thank you for the comment, that's a good idea indeed; will update the code.

I think it's okay, I can handle the changes, just let me know in the comments if you see something that you think should be improved

Reply:
What do you think of the current version ?

Reply:
Left some minor comments but it looks clean, thanks!